packages/lib/services/searchengine/SearchEngine.ts

```diff
@@ -554,6 +554,94 @@ export default class SearchEngine {
 		return SearchEngine.SEARCH_TYPE_FTS;
 	}
 
+	makeSearchTree(str: string) {
+		// (tag:A (any:1 tag:B tag:C)) -> [ "tag:A", [ "any:1", "tag:B", "tag:C"]]
+		let i = 0;
+		function main(): any {
+			const arr = [];
+			let startIndex = i;
+			function addWord() {
+				if (i - 1 > startIndex) {
+					arr.push(str.slice(startIndex, i - 1));
+				}
+			}
+			while (i < str.length) {
+				switch (str[i++]) {
+				case ' ':
+					addWord();
+					startIndex = i;
+					continue;
+				case '(':
+					arr.push(main());
+					startIndex = i;
+					continue;
+				case ')':
+					addWord();
+					return arr;
+				}
+			}
+			addWord();
+			return arr;
+		}
+		return main();
+	}
+
+
+
+	async makeQuery(parseTree: any) {
+
+		const getQueryWithParms = async (searchString: string) => {
+			const parsedQuery = await this.parseQuery(searchString);
+			const { query, params } = queryBuilder(parsedQuery.allTerms);
+			return { query, params };
+		};
+
+		const intersection = (queries: any) => {
+			const result = queries.reduce((acc: any, curr: any, idx: number) => {
+				return {
+					query: `SELECT * FROM (${acc.query}) t${idx - 1} INNER JOIN (${curr.query}) t${idx} ON t${idx - 1}.id=t${idx}.id`,
+					params: acc.params.concat(curr.params),
+				};
+			});
+			return result;
+		};
+
+		const union = (queries: any) => {
+			const query = `SELECT * FROM (${queries.map((x: any) => x.query).join(' UNION ')})`;
+			const params = queries.map((x: any)=>x.params).flat();
+			return { query, params };
+		};
+
+		const makeQueryForNested = async (parseTree: any) => {
+			const nested = parseTree.filter((x: any) => Array.isArray(x));
+			const promises = nested.map((x: any) => {
+				return this.makeQuery(x);
+			});
+			const result = await Promise.all(promises);
+			return result;
+		};
+
+		const makeQueryForTopLevel = async (searchString: any) => {
```

Comment:
Please use an interface. If it's a plain string, then simply set it to "string". Also define a type for the returned data.